axiom/nr_log_events.c

```diff
@@ -8,32 +8,214 @@
+
+  return nr_log_event_priority_comparator((const nr_log_event_t*)a,
+                                          (const nr_log_event_t*)b);
+}
+
+/* All log events popping out of the segment heap go through this
+ * function.
+ *
+ * The log event is discarded (thus removed from the log event heap).
+ */
+static void nr_log_event_discard_wrapper(nr_log_event_t* event,
+                                         void* userdata NRUNUSED) {
+  if (NULL == event) {
+    return;
+  }
+
+  nr_log_event_destroy(&event);
+}
+
+nr_log_events_t* nr_log_events_create(size_t max_events) {
+  nr_log_events_t* events;
+
+  if (NR_MAX_LOG_EVENTS_MAX_SAMPLES_STORED < max_events) {
+    return NULL;
+  }
+
+  events = (nr_log_events_t*)nr_zalloc(sizeof(nr_log_events_t));
+  events->events_allocated = max_events;
+  events->events_used = 0;
+  events->events_seen = 0;
+
+  if (0 != max_events) {
+    events->events = nr_minmax_heap_create(
+        max_events, nr_log_event_wrapped_priority_comparator, NULL,
+        (nr_minmax_heap_dtor_t)nr_log_event_discard_wrapper, NULL);
+  } else {
+    events->events = NULL;
+  }
+
+  return events;
+}
 
+void nr_log_events_destroy(nr_log_events_t** events_ptr) {
+  nr_log_events_t* events;
+
+  if (NULL == events_ptr) {
+    return;
+  }
+  events = *events_ptr;
   if (NULL == events) {
-    return false;
+    return;
+  }
+
+  if (NULL != events->events) {
+    nr_minmax_heap_destroy(&events->events);
   }
+
+  nr_realfree((void**)events_ptr);
+}
+
+size_t nr_log_events_max_events(const nr_log_events_t* events) {
+  if (NULL == events)
+    return 0;
+
+  return events->events_allocated;
+}
+
+size_t nr_log_events_number_seen(const nr_log_events_t* events) {
+  if (NULL == events)
+    return 0;
+
+  return events->events_seen;
+}
+
+size_t nr_log_events_number_saved(const nr_log_events_t* events) {
+  if (NULL == events)
+    return 0;
+
+  return events->events_used;
+}
+
+bool nr_log_events_add_event(nr_log_events_t* events,
+                             const nr_log_event_t* event) {
+  bool events_sampled = false;
+  nr_log_event_t* dup_event;
+
   if (NULL == event) {
     return false;
   }
 
-  events_sampled = nr_analytics_events_is_sampling(events);
+  // need to increment this if a valid event was sent
+  // so take care if it here then sanity check other args
+  if (NULL != events)
+    events->events_seen++;
 
-  log_event_json = nr_log_event_to_json(event);
-  log_event = nr_analytics_event_create_from_string(log_event_json);
-  nr_free(log_event_json);
+  // if no event queue exists or size is 0 then event will be dropped
+  if (NULL == events || NULL == events->events
+      || 0 == events->events_allocated) {
+    return true;
+  }
 
-  nr_analytics_events_add_event(events, log_event, rnd);
-  nr_free(log_event);
+  events_sampled = nr_log_events_is_sampling(events);
+
+  // make a copy of the event so the heap can control its entire life cycle
+  dup_event = nr_log_event_clone(event);
```

Comment:
It seems expensive to clone every log event.  What is the advantage?
Why not just pass control of the original event to the heap and not destroy the event [here](https://github.com/newrelic/newrelic-php-agent/pull/515/files#diff-aff431af9f447c5431ea6466c518a37c44b2491884a829f6c9e4675082ca8186R3385)?

Reply:
The initial implementation (prior to switching to minmax heap) worked where the caller was freeing the event so I modeled this to follow that flow.
Let me see how invasive it would be to change the existing code to not free the event.

Reply:
I believe commit [08a36da](https://github.com/newrelic/newrelic-php-agent/pull/515/commits/08a36dae321abfef2d4b572edfc5ab6af5a07d7d) implements this functionality without the clone.
Please check this commit for correctness.  It does pass axiom-valgrind but is a departure from how the code was originally written so extra review is appreciated.